src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -278,8 +279,12 @@ instance Eval f => Eval (BackendSyntax f) where
       guardFail (eval env val) SemEffectDefer
     Accessor lhs accessor ->
       evalAccessor env (eval env lhs) accessor
-    Update lhs updates ->
-      evalUpdate (eval env lhs) (map (eval env) <$> updates)
+    Update lhs updates -> do
+      let
+        folder acc = case _ of
+          List.Nil -> evalUpdate (eval env lhs) acc
+          List.Cons (Prop key sem) t -> guardFloatLet (eval env sem) \v -> folder (acc <> [ Prop key v ]) t
+      folder [] (List.fromFoldable updates)
```

Comment:
We had talked about making "smart" eval constructors for these, so that we can use them in other evaluation routines and preserve the let floating invariant.